src/QuadForm/Herm/GenusRep.jl

```diff
@@ -48,86 +8,171 @@ end
+
+  if !isdefinite(L)
+    return false, 1*S, bad
+  end
+
+  # TODO: This does not find the prime ideal with smallest norm,
+  # but with smallest minimum ...
+
+  m = rank(L)
+  p = 1
+  P = ideal_type(R)[]
+  while length(P) == 0
+    p = next_prime(p)
+    lp = ideal_type(R)[ p[1] for p in prime_decomposition(R, p)]
+    P = setdiff(lp, bad)
+    if m == 2
+      P = filter(p -> isisotropic(L, p), P)
     end
-    LL = vcat(LL, N, inter)
   end
-  @assert length(LL) == prod(Int[g[2] for g in gens if !definite || g[1] != P0])
-  @assert all(X -> genus(X) == genus(L), LL)
+  Q = prime_decomposition(S, P[1])[1][1]
 
-  if use_mass
-    mass = Hecke.mass(L)
+  if isempty(bad)
+    I = 1 * S
   else
-    mass = zero(fmpq)
+    I = prod(bad) * S
   end
+  n = absolute_norm(Q)
+  if n >= 1000
+    PP = prime_ideals_up_to(S, 1000)
+    for QQ in PP
+      if !iscoprime(QQ, I)
+        continue
+      end
 
-  missing_mass = Ref(mass)
+      if isisotropic(L, QQ)
+        return true, QQ, bad
+      end
+    end
+  end
+  PP = prime_ideals_up_to(S, n)
+  for QQ in PP
+    if !iscoprime(QQ, I)
+      continue
+    end
+    if isisotropic(L, QQ)
+      return true, QQ, bad
+    end
+  end
+  throw(error("Impossible"))
+end
 
-  local result::Vector{typeof(L)}
+function _neighbour(L, B, xG, x, h, P, CC, split)
+  R = order(P)
+  K = nf(R)
+  n = nrows(B)
+  a = involution(L)
 
-  if definite
-    result = typeof(L)[]
-    for L in LL
-      # Should never happen!
-      @assert all(X -> !isisometric(X, L), result)
-      neig = iterated_neighbours(L, P0, max = max, use_auto = use_auto, missing_mass = missing_mass)
-      append!(result, neig)
-      max = max - length(result)
+  local C::ideal_type(R)
+
+  if CC isa Int
+    C = split ? a(P) : P
+  else
+    C = CC
+  end
+
+  I = Int[ i for i in 1:n if !iszero(xG[i])]
+  i = I[1]
+  M = zero_matrix(K, n - length(I), ncols(B))
+  for (k, nk) in enumerate(setdiff(1:n, I))
+    for j in 1:ncols(B)
+      M[k, j] = B[nk, j]
     end
-    for i in 1:length(result)
-      for j in 1:i-1
-        @hassert :GenRep 1 !(isisometric(result[i], result[j])[1])
-      end
+  end
+  CI = fractional_ideal_type(R)[ K(1) * R for j in 1:(n - length(I)) ]
+  for j in I
+    if j != i
+      M = vcat(M, matrix(K, 1, ncols(B), elem_type(K)[B[j, k] - a(h\(divexact(xG[j], xG[i]))) * B[i, k] for k in 1:ncols(B)]))
+      push!(CI, K(1) * R)
     end
+  end
+  M = vcat(M, sub(B, i:i, 1:ncols(B)))
+  push!(CI, fractional_ideal(order(P), P))
+  M = vcat(M, matrix(K, 1, ncols(B), x))
+  push!(CI, inv(C))
+  pm = pseudo_matrix(M, CI)
+  _M = _sum_modules(L, _module_scale_ideal((split ? P : P * C), pseudo_matrix(L)), pm)
+  LL = lattice(ambient_space(L), _M)
+
+  @assert islocally_isometric(L, LL, P)
+  return LL
+end
+
+function stdcallback(list, L)
+  keep = all(LL -> !isisometric(LL,L)[1], list)
+  return keep, true
+end
+
+#TODO : implement isometry check for indefinite lattice in the same space
+function relcallback(list,L)
```

Comment:
Did we always have this? I wonder what the "rel" stands for?

Reply:
No, only for `neighbours_with_ppower` because it is called outside of the loop running for definite lattices. So in particular, there is this test running even for indefinite lattices.  M. Kirschmer seems to avoid this by not allowing the use of the automorphism group but even though we do the same, asking for an isometry test will use this group... Here the "rel" stands for "relative", that's a way I found to differentiate it from `stdcallback` for which I guess "std" stands for "standard".

Reply:
Ah and maybe I misunderstood: we never had this function, I added it waiting to fix isometry check for indefinite lattices

Reply:
OK

Reply:
Using the automorphism group does not make sense in the indefinite case. There we do not need to iterate over all the p-neighbors up to isometry. Instead we just visit all spinor genera.

Reply:
In this context it is not directly used, but the short vectors and Cholesky decomposition to check if two lattices are isometric... because when adding a new candidates as a neighbour, we need to check if it was not already in the list of candidates already computed